File: ds/python/queue/queue.py

class Queue:
    class __Node:
        def __init__(self, data):
            self.data = data
            self.next = None

    def __init__(self):
        self.front = None
        self.rear = None

    def __iter__(self):
        temp = self.front
        while temp:
            yield temp.data
            temp = temp.next

    def __repr__(self):
        return f"Queue({', '.join(str(node) for node in self)})"

    def __len__(self):
        return len(tuple(iter(self)))

    @classmethod
    def getNewNode(cls, data):
        return cls.__Node(data)

    def enqueue(self, node):
        if self.rear is None:
            self.rear = node
            self.front = node
        else:
            self.rear.next = node
            self.rear = node

    def dequeue(self):
        if self.isEmpty():
            raise Exception("Queue is Empty")

        front = self.front
        self.front = self.front.next
        return front.data

    def getFront(self):
        if self.isEmpty():
            raise Exception("Queue is Empty")

        return self.front.data

    def getRear(self):
        if self.isEmpty():
            raise Exception("Queue is Empty")

        return self.rear.data

    def isEmpty(self):
        return self.front is None or self.rear is None


def main():
    queue = Queue()
    queue.enqueue(Queue.getNewNode("Eenne"))
    print(f"{queue}\nLength: {len(queue)}")
    queue.enqueue(Queue.getNewNode("Menne"))
    print(f"{queue}\nLength: {len(queue)}")
    queue.enqueue(Queue.getNewNode("Manne"))
    print(f"{queue}\nLength: {len(queue)}")
    print(f"\n{queue.dequeue()} removed from queue")
    print(f"{queue}\nLength: {len(queue)}")
    print(queue.getFront(), " at the front of queue")
    print(queue.getRear(), " at the back of queue")

if __name__ == "__main__":
    main()

File: ds/C/Queue/queue.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct Node Node;
struct Node {
    int data;
    Node* next;
};

Node* front = NULL;
Node* rear = NULL;

bool isEmpty() {
    return (front == NULL && rear == NULL);
}

int Front() {
    if (isEmpty()) return -1;
    return front->data;
}

void enqueue(int x) {
    Node* newNode = malloc(sizeof(Node));
    newNode->data = x;
    newNode->next = NULL;

    if (isEmpty()) front = rear = newNode;
    else {
        rear->next = newNode;
        rear = newNode;
    }
}

void dequeue() {
    Node* tmp = front;
    if (isEmpty()) return;
    else if (front == rear) front = rear = NULL;
    else front = front->next;
    free(tmp);
}

void print() {
    if (isEmpty()) printf("The list is empty.\n");
    else {
        Node* tmp = front;
        while (tmp != NULL) {
            printf("%d ", tmp->data);
            tmp = tmp->next;
        }
        printf("\n");
    }
}

int main(void) {
    enqueue(5);print();
    enqueue(6);print();
    enqueue(7);print();
    dequeue(); print();
    enqueue(8);print();
    dequeue(); print();
    return 0;
}
